### PR TITLE
Model links lazy evaluation in pipeline code

### PR DESCRIPTION
**Pull request by avishniakov**
## Describe changes
I implemented the ability to do lazy loading of Artifact Version, Artifact Version Metadata, and Model Version Metadata inside the pipeline body. The following syntax will evaluate objects only during pipeline execution (before it was fixed on pipeline compilation):
```python
    @pipeline(model_version=ModelVersion(name=model_name),...)
    def my_pipe():
        model_version = get_pipeline_context().model_version
        artifact = model_version.get_artifact("artifact_name")
        artifact_metadata = artifact.run_metadata["metadata_key"]
        model_metadata = model_version.run_metadata["metadata_key2"]
        my_step(artifact, artifact_metadata, model_metadata)
```

**Side changes:**
- `log_model_version_metadata` is now listed on root level (`from zenml import log_model_version_metadata`)

**Side breaking change:**
- Unified `metdata` and `run_metadata` across Artifact and Model Versions:`ModelVersion().metadata` becomes `ModelVersion().run_metadata`

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] I have based my new branch on `develop` and the open PR is targeting `develop`. If your branch wasn't based on develop read [Contribution guide on rebasing branch to develop](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md#-pull-requests-rebase-your-branch-on-develop).
- [x] If my changes require changes to the dashboard, these changes are communicated/requested.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: .github/workflows/update-templates-to-examples.yml [hunk 1]
```diff
@@ -95,7 +95,7 @@ jobs:
         with:
           github-token: ${{ secrets.GITHUB_TOKEN }}
           script: |
-            github.issues.createComment({
+            github.rest.issues.createComment({
```
Comment (strickvl) [inline .github/workflows/update-templates-to-examples.yml:98]: their API changed?

File: src/zenml/config/pipeline_configurations.py [hunk 2]
```diff
@@ -86,3 +89,23 @@ def docker_settings(self) -> "DockerSettings":
             DOCKER_SETTINGS_KEY, {}
         )
         return DockerSettings.parse_obj(model_or_dict)
+
+    @contextmanager
+    def in_design_time(self) -> Iterator[None]:
+        """Context manager to mark a pipeline as being in design time.
```
Comment (strickvl) [inline src/zenml/config/pipeline_configurations.py:95]: What does something being 'in design time' mean? Maybe add something to the docstring on this?

Comment (strickvl) [reply src/zenml/config/pipeline_configurations.py:95]: I read a bit and it's 'a thing'! Is this what you also have elsewhere referred to as 'composition'?

File: src/zenml/lazy_load/model_version.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
+#  Copyright (c) ZenML GmbH 2023. All Rights Reserved.
```
Comment (strickvl) [inline src/zenml/lazy_load/model_version.py:1]:
```suggestion
#  Copyright (c) ZenML GmbH 2024. All Rights Reserved.
```

File: src/zenml/lazy_load/model_version.py
Comment (strickvl) [inline src/zenml/lazy_load/model_version.py:1]: I defer to your experience, but is a whole new `lazy_load` model where this stuff should live?

File: src/zenml/lazy_load/run_metadata.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
+#  Copyright (c) ZenML GmbH 2023. All Rights Reserved.
```
Comment (strickvl) [inline src/zenml/lazy_load/run_metadata.py:1]:
```suggestion
#  Copyright (c) ZenML GmbH 2024. All Rights Reserved.
```

File: src/zenml/models/v2/base/internal.py [hunk 5]
```diff
@@ -15,10 +15,13 @@
 
 from typing import Type, TypeVar
 
+from zenml.logger import get_logger
 from zenml.models.v2.base.base import BaseRequest
 
 T = TypeVar("T", bound="BaseRequest")
 
+logger = get_logger(__name__)
```
Comment (strickvl) [inline src/zenml/models/v2/base/internal.py:23]: why was this added? I guess for debugging? Maybe remove?

File: tests/integration/functional/pipelines/test_pipeline_context.py [hunk 6]
```diff
@@ -97,3 +100,46 @@ def test_that_argument_as_get_artifact_of_model_version_in_pipeline_context_fail
...
+    return "artifact_data_" + ver
+
+
+@step
+def asserter(artifact: str, artifact_metadata: str, model_metadata: str):
+    """Assert that passed in values are loaded in lazy mode, since they should not exists on compose run."""
+    ver = get_step_context().model_version.version
+    assert artifact == "artifact_data_" + ver
+    assert artifact_metadata == "artifact_meta_" + ver
+    assert model_metadata == "model_meta_" + ver
```
Comment (strickvl) [inline tests/integration/functional/pipelines/test_pipeline_context.py:120]: is the idea that we're going to use these more in the future / going forward elsewhere?

**Review by strickvl — CHANGES_REQUESTED** [review]

File: .github/workflows/update-templates-to-examples.yml (hunk 1 above)
Comment (avishniakov) [reply .github/workflows/update-templates-to-examples.yml:98]: Yeah, since your bump of GH action versions old endpoint was not working anymore

File: src/zenml/config/pipeline_configurations.py (hunk 2 above)
Comment (avishniakov) [reply src/zenml/config/pipeline_configurations.py:95]: Yep, design time is a composition in my terms. Shall I unify it all to composition/execution instead of design/run times?

Comment (strickvl) [reply src/zenml/config/pipeline_configurations.py:95]: I don't mind really, but yeah it should be one word. I heard them both first this week so not even sure it's a thing in ZenML docs?

File: tests/integration/functional/pipelines/test_pipeline_context.py (hunk 6 above)
Comment (avishniakov) [reply tests/integration/functional/pipelines/test_pipeline_context.py:120]: Not necessarily, but steps cannot be defined as inner functions, cause it messes up with the way we refer to the sources on step execution later on.

File: src/zenml/models/v2/base/internal.py (hunk 5 above)
Comment (avishniakov) [reply src/zenml/models/v2/base/internal.py:23]: Yeah, will remove this

File: tests/integration/functional/pipelines/test_pipeline_context.py (hunk 6 above)
Comment (strickvl) [reply tests/integration/functional/pipelines/test_pipeline_context.py:120]: Ah right yes. got it.

File: src/zenml/config/pipeline_configurations.py (hunk 2 above)
Comment (avishniakov) [reply src/zenml/config/pipeline_configurations.py:95]: Seems that `compile` is the term used internally for design time. Will rename

Comment (avishniakov) [reply src/zenml/config/pipeline_configurations.py]: Renamed

File: src/zenml/models/v2/base/internal.py (hunk 5 above)
Comment (avishniakov) [reply src/zenml/models/v2/base/internal.py]: Removed

File: src/zenml/lazy_load/model_version.py
Comment (avishniakov) [reply src/zenml/lazy_load/model_version.py:1]: Good question, I don't remember the real motivation behind it (might be due to some circular imports). I reworked it now, let's see if tests are unaffected.

**Review by strickvl — APPROVED** [review]
🙏🏻

File: src/zenml/metadata/lazy_load.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        Args:
+            key: The metadata key.
+
+        Returns:
+            The metadata lazy loader wrapper for the given key.
+        """
+        from zenml.models import RunMetadataResponse
+
+        return RunMetadataResponse(
+            id=uuid4(),
```
Comment (schustmi) [inline src/zenml/metadata/lazy_load.py:61]: As you already pointed out in the PR description, not a huge fan of this. I don't think we can easily remove that from the `BaseResponse` class (for our actual responses, we do always want the ID to be in there). There's also a problem when users in their pipeline function try to access any properties of the lazy responses, they will get random values or strange error messages.

At least the error messages we could prevent by maybe having a separate subclass for it, but not convinced this is a good idea..
```python
class LazyArtifactVersionResponse(ArtifactVersionResponse):
     _lazy_load_name: Optional[str] = None
    _lazy_load_version: Optional[str] = None
    _lazy_load_model_version: Optional["ModelVersion"] = None

    # We should probably do this for all properties, otherwise users will run into strange error message when they 
    # try to access any of those on a lazy object
    @property
    def id(self) -> UUID:
        raise RuntimeError("ID can't be accessed for lazy AVR")

    ...
```

Comment (schustmi) [reply src/zenml/metadata/lazy_load.py:61]: Actually instead of overwriting all properties, we probably only need to overwrite the `get_body()` and `get_metadata()` functions to raise errors which describe the issue.

**Review by schustmi — CHANGES_REQUESTED** [review]

File: src/zenml/new/pipelines/pipeline_context.py [hunk 8]
```diff
@@ -108,3 +108,4 @@ def __init__(self, pipeline_configuration: "PipelineConfiguration"):
         self.settings = pipeline_configuration.settings
         self.extra = pipeline_configuration.extra
         self.model_version = pipeline_configuration.model_version
+        self.is_compiling = pipeline_configuration.is_compiling
```
[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/new/pipelines/pipeline_context.py:111]: Very valid, thanks!

File: src/zenml/steps/base_step.py [hunk 9]
```diff
@@ -495,6 +500,26 @@ def _parse_call_args(
...
+            elif isinstance(value, ArtifactVersionResponse):
+                if value._lazy_load_model_version:
+                    model_artifacts_or_metadata[
+                        key
+                    ] = ModelVersionDataLazyLoader(
+                        model_version=value._lazy_load_model_version,
+                        artifact_name=value._lazy_load_name,
+                        artifact_version=value._lazy_load_version,
+                        metadata_name=None,
+                    )
```
[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/steps/base_step.py:512]: No more if else after rework

File: src/zenml/config/pipeline_configurations.py [hunk 10]
```diff
@@ -53,6 +54,8 @@ class PipelineConfiguration(PipelineConfigurationUpdate):
 
     name: str
 
+    _compiling_now: List[bool] = [False]
```
[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/config/pipeline_configurations.py:57]: It was a workaround to make strict immutable model to be mutable, not needed anymore

File: src/zenml/metadata/lazy_load.py (hunk 7 above)
Comment (avishniakov) [reply src/zenml/metadata/lazy_load.py:61]: Super, thanks!

File: src/zenml/orchestrators/input_utils.py [hunk 11]
```diff
@@ -81,6 +83,43 @@ def resolve_step_inputs(
...
+            if artifact_ := config_.model_version.get_artifact(
+                config_.artifact_name, config_.artifact_version
+            ):
+                input_artifacts[name] = artifact_
+            else:
+                issue_found = True
+        elif config_.artifact_name is None and config_.metadata_name:
+            # metadata values should go directly in parameters, as primitive types
+            step.config.parameters[name] = config_.model_version.run_metadata[
+                config_.metadata_name
```
[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/orchestrators/input_utils.py]: Added catcher

**Comment by avishniakov** [comment]
Thanks a lot @schustmi - a golden review 👍🏼 You solved all design issues with one shot. Please read it again once CI passes.

**Review by stefannica — APPROVED** [review]
I agree to the overall solution of passing references to model versions and artifacts at pipeline compilation time and deferring their resolution at pipeline execution time. I can't say I fully understand all the implications of this, given that I'm not familiar with this area of code. Good to go from my end.

**Review by schustmi — APPROVED** [review]
